src/Pinch/Transport.hs

```diff
@@ -47,28 +47,23 @@ data Transport
 
 -- | Creates a thrift framed transport. See also <https://github.com/apache/thrift/blob/master/doc/specs/thrift-rpc.md#framed-vs-unframed-transport>.
 framedTransport :: Connection c => c -> IO Transport
-framedTransport c = pure $ Transport writeMsg readMsg where
+framedTransport c = do
+  readBuffer <- newIORef mempty
+  pure $ Transport writeMsg (readMsg readBuffer) where
   writeMsg msg = do
     cPut c $ B.runBuilder $ B.int32BE (fromIntegral $ B.getSize msg)
     cPut c $ B.runBuilder msg
 
-  readMsg p = do
-    szBs <- getExactly c 4
-    if BS.length szBs < 4
-      then
-        pure $ RREOF
-      else do
-        let sz = fromIntegral <$> G.runGet G.getInt32be szBs
-        case sz of
-          Right x -> do
-            msgBs <- getExactly c x
-            pure $ if BS.length msgBs < x
-              then
-                -- less data has been returned than expected. This means we have reached EOF.
-                RREOF
-              else
-                either RRFailure RRSuccess $ G.runGet p msgBs
-          Left s -> pure $ RRFailure $ "Invalid frame size: " ++ show s
+  readMsg readBuffer parser = do
+    let 
+      frameParser = do 
+        size <- G.getInt32be
+        G.isolate (fromIntegral size) parser
+    
+    initial <- readIORef readBuffer
+    (leftovers, r) <- runGetWith (cGetSome c) frameParser initial
```

Comment:
I guess this works out fine here if `initial` is empty because in runGetWith we will get a `Partial` parse result and automatically try to fetch more data?

Reply:
Yes, exactly. It's fine for the initial chunk to be empty. It's only once it started parsing and you pass an empty chunk it'd be regarded as EOF.